url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/layout/JGraphLayoutExample.java

```diff
@@ -0,0 +1,350 @@
+		boolean notEmpty = (graph.getModel().getRootCount() > 0);
+		// hide.setEnabled(enabled);
+		expand.setEnabled(enabled);
+		expandAll.setEnabled(enabled);
+		collapse.setEnabled(enabled);
+		configure.setEnabled(layoutCache != null && layoutCache.layout != null);
+		if (layoutActions != null) {
+			for (int i = 0; i < layoutActions.length; i++) {
+				layoutActions[i].setEnabled(notEmpty);
+			}
+		}
+	}
+
+	protected void configureEncoder(XMLEncoder encoder) {
+		super.configureEncoder(encoder);
+		encoder
+				.setPersistenceDelegate(JGraphExampleLayoutCache.class,
+						new DefaultPersistenceDelegate(new String[] { "model",
+								"factory", "cellViews", "hiddenCellViews",
+								"partial" }));
+		encoder.setPersistenceDelegate(JGraphExampleGraph.class,
+				new DefaultPersistenceDelegate(new String[] { "model",
+						"graphLayoutCache" }));
+		encoder.setPersistenceDelegate(JGraphParallelRouter.class,
+				new PersistenceDelegate() {
+					protected Expression instantiate(Object oldInstance,
+							Encoder out) {
+						return new Expression(oldInstance,
+								JGraphParallelRouter.class,
+								"getSharedInstance", null);
+					}
+				});
+	}
+
+	/**
+	 * Hook from GraphEd to add action button to the tool bar
+	 */
+	public JToolBar createToolBar() {
+		JToolBar toolbar = super.createToolBar();
+
+		// Configure layout
+		configure = new AbstractAction() {
+			public void actionPerformed(ActionEvent e) {
+				layoutCache.configureLayout();
+			}
+		};
+		URL url = getClass().getClassLoader().getResource(
+				"com/jgraph/layout/image/configure.gif");
+		configure.putValue(Action.SMALL_ICON, new ImageIcon(url));
+		configure.setEnabled(false);
+		toolbar.add(configure);
+
+		// AutoLayout toggle
+		autoLayoutButton = new JToggleButton();
+		autoLayoutButton.setSelected(false);
+		url = getClass().getClassLoader().getResource(
+				"com/jgraph/layout/image/layout.gif");
+		autoLayoutButton.setIcon(new ImageIcon(url));
+		autoLayoutButton.addActionListener(new ActionListener() {
+			public void actionPerformed(ActionEvent e) {
+				layoutCache.autolayout = autoLayoutButton.isSelected();
+			}
+		});
+		toolbar.add(autoLayoutButton);
+
+		// JButton button = new JButton("My");
+		// toolbar.add(button);
+		// button.addActionListener(new ActionListener() {
+		//
+		// public void actionPerformed(ActionEvent event) {
+		// // Add your code here
+		// }
+		//
+		// });
+		return toolbar;
+	}
+
+	public void openFile() {
+		int returnValue = JFileChooser.CANCEL_OPTION;
+		initFileChooser();
+		returnValue = fileChooser.showOpenDialog(graph);
+		if (returnValue == JFileChooser.APPROVE_OPTION) {
+			Container parent = graph.getParent();
+			BasicMarqueeHandler marqueeHandler = graph.getMarqueeHandler();
+			try {
+				uninstallListeners(graph);
+				parent.remove(graph);
+				XMLDecoder decoder = new XMLDecoder(new BufferedInputStream(
+						new FileInputStream(fileChooser.getSelectedFile())));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *BufferedInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 289 evaluates to true. Ensure the object returned by *BufferedInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).